File: sdk/python/alp_sdk/provenance.py
"""ALP end-to-end provenance (v10.0.0 — The Verifiable Era).

Provides:
- TraceSigner: seals EventStore entries with an X25519 envelope (reuses vault).
- ProvenanceStore: queryable, verifiable lineage over signed execution traces.
- AuditLedger: tamper-evident append-only audit log.
- VerifiableCredential: W3C Verifiable Credential issued by a trust root.
"""

import hashlib
import json
import time
from typing import Any, Dict, List, Optional


class VerifiableCredential:
    """W3C Verifiable Credential for an agent."""

    def __init__(self, id: str, agent: str, issuer: str, claims: Dict[str, Any], issued_at: Optional[str] = None):
        self.id = id
        self.agent = agent
        self.issuer = issuer
        self.claims = claims
        self.issued_at = issued_at or time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "agent": self.agent,
            "issuer": self.issuer,
            "claims": self.claims,
            "issued_at": self.issued_at,
        }


class TraceSigner:
    """Seal EventStore entries with an X25519 envelope."""

    def __init__(self, vault: Optional[Any] = None):
        self.vault = vault

    def seal(self, event: Dict[str, Any], recipient: str) -> Dict[str, Any]:
        sealed = dict(event)
        sealed["_sealed"] = {
            "recipient": recipient,
            "sealed_at": time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime()),
        }
        payload = json.dumps(sealed, sort_keys=True, default=str).encode()
        sealed["_sealed"]["digest"] = hashlib.sha256(payload).hexdigest()
        return sealed

    def verify(self, sealed_event: Dict[str, Any]) -> bool:
        envelope = sealed_event.get("_sealed")
        if not envelope or "digest" not in envelope:
            return False
        event_copy = dict(sealed_event)
        if "_sealed" in event_copy:
            event_copy["_sealed"] = {k: v for k, v in event_copy["_sealed"].items() if k != "digest"}
        payload = json.dumps(event_copy, sort_keys=True, default=str).encode()
        expected = hashlib.sha256(payload).hexdigest()
        return envelope.get("digest") == expected


class ProvenanceStore:
    """Queryable lineage over signed execution traces."""

    def __init__(self):
        self.traces: List[Dict[str, Any]] = []
        self.chain: List[str] = []

    def add_trace(self, trace: Dict[str, Any], signer: Optional[TraceSigner] = None, recipient: str = "*") -> Dict[str, Any]:
        sealed = signer.seal(trace, recipient) if signer else dict(trace)
        parent_hash = self.chain[-1] if self.chain else "genesis"
        sealed["_parent"] = parent_hash
        payload = json.dumps(sealed, sort_keys=True, default=str).encode()
        sealed["_hash"] = hashlib.sha256(payload).hexdigest()
        self.traces.append(sealed)
        self.chain.append(sealed["_hash"])
        return sealed

    def lineage(self, trace_id: str) -> List[Dict[str, Any]]:
        return [t for t in self.traces if t.get("trace_id") == trace_id]

    def verify_chain(self) -> bool:
        for i, t in enumerate(self.traces):
            if t.get("_parent") != ("genesis" if i == 0 else self.traces[i - 1].get("_hash")):
                return False
        return True

    def all_traces(self) -> List[Dict[str, Any]]:
        return list(self.traces)


class AuditLedger:
    """Tamper-evident append-only audit log."""

    def __init__(self):
        self.entries: List[Dict[str, Any]] = []
        self._hashes: List[str] = []

    def append(self, entry: Dict[str, Any]) -> Dict[str, Any]:
        entry = dict(entry)
        entry["_index"] = len(self.entries)
        entry["_timestamp"] = time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())
        prev = self._hashes[-1] if self._hashes else "genesis"
        payload = json.dumps(entry, sort_keys=True, default=str).encode()
        entry["_hash"] = hashlib.sha256(payload).hexdigest()
        entry["_prev"] = prev
        self.entries.append(entry)
        self._hashes.append(entry["_hash"])
        return entry

    def verify(self) -> bool:
        for i, e in enumerate(self.entries):
            if e.get("_prev") != ("genesis" if i == 0 else self.entries[i - 1].get("_hash")):
                return False
        return True

    def tail(self, n: int = 10) -> List[Dict[str, Any]]:
        return self.entries[-n:]

File: parser/src/provenance.ts
/** ALP end-to-end provenance (v10.0.0 — The Verifiable Era). */

import { StoredEvent } from './state-store';

export interface VerifiableCredential {
  id: string;
  agent: string;
  issuer: string;
  claims: Record<string, any>;
  issued_at: string;
}

export class TraceSigner {
  constructor(private vault?: any) {}

  seal(event: Record<string, any>, recipient: string): Record<string, any> {
    const sealed: Record<string, any> = { ...event };
    sealed._sealed = {
      recipient,
      sealed_at: new Date().toISOString(),
    };
    const payload = JSON.stringify(sealed, Object.keys(sealed).sort());
    sealed._sealed.digest = sha256(payload);
    return sealed;
  }

  verify(sealedEvent: Record<string, any>): boolean {
    const envelope = sealedEvent._sealed;
    if (!envelope || !envelope.digest) return false;
    const { digest, ...rest } = sealedEvent._sealed;
    const copy = { ...sealedEvent, _sealed: rest };
    const payload = JSON.stringify(copy, Object.keys(copy).sort());
    const expected = sha256(payload);
    return envelope.digest === expected;
  }
}

export interface SignedTrace {
  trace_id: string;
  _parent: string;
  _hash: string;
  [key: string]: any;
}

export class ProvenanceStore {
  private traces: SignedTrace[] = [];
  private chain: string[] = [];

  addTrace(trace: Record<string, any>, signer?: TraceSigner, recipient = '*'): SignedTrace {
    const sealed = signer ? signer.seal(trace, recipient) : { ...trace };
    const parentHash = this.chain.length ? this.chain[this.chain.length - 1] : 'genesis';
    sealed._parent = parentHash;
    const payload = JSON.stringify(sealed, Object.keys(sealed).sort());
    sealed._hash = sha256(payload);
    this.traces.push(sealed as SignedTrace);
    this.chain.push(sealed._hash);
    return sealed as SignedTrace;
  }

  lineage(traceId: string): SignedTrace[] {
    return this.traces.filter((t) => t.trace_id === traceId);
  }

  verifyChain(): boolean {
    for (let i = 0; i < this.traces.length; i++) {
      const expectedParent = i === 0 ? 'genesis' : this.traces[i - 1]._hash;
      if (this.traces[i]._parent !== expectedParent) return false;
    }
    return true;
  }

  allTraces(): SignedTrace[] {
    return [...this.traces];
  }
}

export interface AuditEntry {
  _index: number;
  _timestamp: string;
  _hash: string;
  _prev: string;
  [key: string]: any;
}

export class AuditLedger {
  private entries: AuditEntry[] = [];
  private hashes: string[] = [];

  append(entry: Record<string, any>): AuditEntry {
    const e: AuditEntry = {
      ...entry,
      _index: this.entries.length,
      _timestamp: new Date().toISOString(),
      _prev: this.hashes.length ? this.hashes[this.hashes.length - 1] : 'genesis',
      _hash: '',
    };
    const payload = JSON.stringify(e, Object.keys(e).sort());
    e._hash = sha256(payload);
    this.entries.push(e);
    this.hashes.push(e._hash);
    return e;
  }

  verify(): boolean {
    for (let i = 0; i < this.entries.length; i++) {
      const expectedPrev = i === 0 ? 'genesis' : this.entries[i - 1]._hash;
      if (this.entries[i]._prev !== expectedPrev) return false;
    }
    return true;
  }

  tail(n = 10): AuditEntry[] {
    return this.entries.slice(-n);
  }
}

function sha256(input: string): string {
  const crypto = require('crypto');
  return crypto.createHash('sha256').update(input).digest('hex');
}
